File: Vachanam/Document/DocumentFormat.swift
//
//  DocumentFormat.swift
//  Vachanam
//
//  Supported document formats for Vachanam multi-format ingestion.
//

import Foundation
import UniformTypeIdentifiers

public enum DocumentFormat: String, CaseIterable, Identifiable, Codable, Sendable {
    case pdf
    case epub
    case markdown
    case plainText
    case webArticle
    
    public var id: String { rawValue }
    
    public var title: String {
        switch self {
        case .pdf: return "PDF Document"
        case .epub: return "EPUB Book"
        case .markdown: return "Markdown Note"
        case .plainText: return "Plain Text"
        case .webArticle: return "Web Article"
        }
    }
    
    public var badgeText: String {
        switch self {
        case .pdf: return "PDF"
        case .epub: return "EPUB"
        case .markdown: return "MD"
        case .plainText: return "TXT"
        case .webArticle: return "WEB"
        }
    }
    
    public var systemImage: String {
        switch self {
        case .pdf: return "doc.text.fill"
        case .epub: return "books.vertical.fill"
        case .markdown: return "text.badge.checkmark"
        case .plainText: return "doc.plaintext.fill"
        case .webArticle: return "globe"
        }
    }
    
    public static func detect(from url: URL) -> DocumentFormat {
        let ext = url.pathExtension.lowercased()
        switch ext {
        case "pdf":
            return .pdf
        case "epub":
            return .epub
        case "md", "markdown":
            return .markdown
        case "txt", "text":
            return .plainText
        default:
            if url.scheme == "http" || url.scheme == "https" {
                return .webArticle
            }
            return .plainText
        }
    }
}

File: VachanamAndroid/app/src/main/java/com/vachanam/reader/data/model/DocumentFormat.kt
package com.vachanam.reader.data.model

import java.io.File
import java.util.Locale

enum class DocumentFormat(val displayName: String, val extensions: List<String>) {
    PDF("PDF Document", listOf("pdf")),
    EPUB("EPUB Book", listOf("epub")),
    MARKDOWN("Markdown", listOf("md", "markdown")),
    PLAIN_TEXT("Plain Text", listOf("txt", "text")),
    WEB_ARTICLE("Web Article", listOf("html", "htm", "web"));

    companion object {
        fun detect(file: File): DocumentFormat {
            val ext = file.extension.lowercase(Locale.ROOT)
            return detect(ext)
        }

        fun detect(extension: String): DocumentFormat {
            val ext = extension.lowercase(Locale.ROOT).removePrefix(".")
            for (fmt in entries) {
                if (fmt.extensions.contains(ext)) {
                    return fmt
                }
            }
            return PLAIN_TEXT
        }

        fun detectFromUrl(url: String): DocumentFormat {
            val lower = url.lowercase(Locale.ROOT)
            return when {
                lower.endsWith(".pdf") -> PDF
                lower.endsWith(".epub") -> EPUB
                lower.endsWith(".md") || lower.endsWith(".markdown") -> MARKDOWN
                lower.endsWith(".txt") -> PLAIN_TEXT
                lower.startsWith("http://") || lower.startsWith("https://") -> WEB_ARTICLE
                else -> PLAIN_TEXT
            }
        }
    }
}
